lago/hooks.py

```diff
@@ -0,0 +1,254 @@
+# coding=utf-8
+import functools
+import os
+from os import path
+import logging
+import shutil
+
+import log_utils
+import utils
+import lockfile
+
+LOGGER = logging.getLogger(__name__)
+LogTask = functools.partial(log_utils.LogTask, logger=LOGGER)
+log_task = functools.partial(log_utils.log_task, logger=LOGGER)
+"""
+Hooks
+======
+Run scripts before or after a Lago command.
+The script will be run on the host which runs Lago.
+"""
+
+
+def with_hooks(func):
+    """
+        Decorate a callable to run with hooks.
+        If without hooks==True, don't run the hooks, just the callable.
+
+        Args:
+            func(callable): callable to decorate
+
+        Returns:
+            The value returned by calling to func
+
+    """
+
+    @functools.wraps(func)
+    def wrap(prefix, without_hooks=False, *args, **kwargs):
+        kwargs['prefix'] = prefix
+
+        if without_hooks:
+            LOGGER.debug('without_hooks=True, skipping hooks')
+            return func(*args, **kwargs)
+
+        cmd = func.__name__
+        if cmd.startswith('do_'):
+            cmd = cmd[3:]
+
+        hooks = Hooks(prefix.paths.hooks())
+        hooks.run_pre_hooks(cmd)
+        result = func(*args, **kwargs)
+        hooks.run_post_hooks(cmd)
```

Comment:
1. Post hook won't run if 'func' throws an exception, do we want to change this behavior ?
2. When passing 'result' to the post hook I need to assume that it is a string, or can be evaluated to a meaningful string, I'm not sure that I can take this assumption.